src/main/java/com/vmware/vipclient/i18n/messages/service/ComponentService.java

```diff
@@ -111,20 +131,43 @@ public MessageCacheItem getMessages(Iterator<Locale> fallbackLocalesIter) {
 		} else { // Item is not in cache.
 			ProductService ps = new ProductService(dto);
 			Locale locale = Locale.forLanguageTag(dto.getLocale());
-			if (ps.isSupportedLocale(locale) || VIPCfg.getInstance().isPseudo()) {
+
+			/*
+			 * Do not block createCacheItem if set of supported locales is not in cache (i.e. ps.getSupportedLocales(false).isEmpty()).
+			 * This happens either when cache is not initialized, OR previous attempts to fetch the set had failed.
+			 */
+			if (ps.getSupportedLocales(false).isEmpty() || ps.isSupportedLocale(false, locale) || VIPCfg.getInstance().isPseudo()) {
 				cacheItem = createCacheItem(); // Fetch for the requested locale from data store, create cacheItem and store in cache
 				if (cacheItem.getCachedData().isEmpty())  // Failed to fetch messages for the requested locale
 					cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
 			} else   // Requested locale is not supported and does not match any supported locales
 				cacheItem = getFallbackLocaleMessages(fallbackLocalesIter);
 		}
+
 		return cacheItem;
 	}
 
+	private void refreshSupportedLocalesTask() {
+		ProductService ps = new ProductService(dto);
+		Runnable runnable = () -> {
+			try {
+				ps.getSupportedLocales();
+			} catch (Exception e) {
+				logger.error("Failed to refresh list of supported locales.");
+			}
+		};
+		new Thread(runnable).start();
+	}
+
 	private void doLocaleMatching() {
-		Locale matchedLocale = LocaleUtility.pickupLocaleFromList(new ProductService(dto).getSupportedLocales(), Locale.forLanguageTag(dto.getLocale()));
-		if (matchedLocale != null) // Requested locale matches a supported locale (eg. requested locale "fr_CA matches supported locale "fr")
+		//Refresh the cache of supported locales as needed in a separate thread (non-blocking).
+		refreshSupportedLocalesTask();
```

Comment:
By our design, the refreshment should be done at each time getting supported locales rather than here.

Reply:
That is not true. See Locale Matching part of the design spec: https://confluence.eng.vmware.com/pages/viewpage.action?pageId=530637182#SingletonClientLibrary-Caching,FallbackMechanism,andOtherFeatures-L10nLocaleMatching
If "List of supported locales in cache" = N, it should return the requested originally requested locale.

In the Query part of the design spec: https://confluence.eng.vmware.com/pages/viewpage.action?pageId=530637182#SingletonClientLibrary-Caching,FallbackMechanism,andOtherFeatures-L10nQuery, you will see that there is no judgment box for "L10n Locale Matching". The path just continues no matter what the result of the locale matching is (non-blocking)